### PR TITLE
Implementing block buildQueryDialog()

### PR DESCRIPTION
**Pull request by hfazai**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 1]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+
+      /* skip fields we don't want to show */
+      if (field.getPriority() == 0) {
+        continue
+      }
+
+      /* skip fields with fixed value */
+      if (!field.isNull(activeRecord) &&
+              field.getSearchOperator() == VConstants.SOP_EQ &&
+              field.getSql(activeRecord)!!.indexOf('*') >= 0) {
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1860]: change to contains & fix code style

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 2]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+        continue
+      }
+      query_tab[query_cnt++] = field
+    }
+
+    /* (bubble) sort fields wrt priorities */
+    for (i in query_cnt - 1 downTo 1) {
+      var swapped = false
+      for (j in 0 until i) {
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1869]: add empty line

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 3]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+        if (abs(query_tab[j]!!.getPriority()) < abs(query_tab[j + 1]!!.getPriority())) {
+          val tmp = query_tab[j]
+          query_tab[j] = query_tab[j + 1]
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1872]: add empty line

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 4]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+
+    /* add the DB column of the ID field. */
+    columns.add(idColumn_)
+
+    /* ... and now their order */
+    var orderbuf = ""
+    var orderSize = 0
+    //val maxCharacters: Int = form.getDBContext().getDefaultConnection().getMaximumCharactersCountInOrderBy() TODO
+    //val maxColumns: Int = form.getDBContext().getDefaultConnection().getMaximumColumnsInOrderBy() TODO
+
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1897]: do we need this code!

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 5]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+    val orderList = mutableListOf<Pair<Column<*>, SortOrder>>()
+
+    for (i in 0 until query_cnt) {
+      // control the size (nbr of columns and size of characters in an "order by" clause)
+      val size = query_tab[i]!!.width * query_tab[i]!!.height
+      /*if (orderSize + size > maxCharacters || i > maxColumns) {
+        break  // TODO
+      }*/
+      orderSize += size
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1906]: we dont use this values later

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 6]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+        orderList.add(columns[i] to SortOrder.DESC)
+      } else {
+        orderList.add(columns[i] to SortOrder.ASC)
+      }
+    }
+
+    /* query from where ? */
+    val tables = getSearchTables_()
+    val conditions = getSearchConditions_()
+
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1918]: delete empty line

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 7]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+    val values = Array(query_cnt) { arrayOfNulls<Any>(fetchSize) }
+    val ids = IntArray(fetchSize)
+    var rows = 0
+
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1922]: do we need to open transaction here!

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 8]
```diff
@@ -1834,8 +1837,119 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+        values[i][rows] = query_tab[i]!!.retrieveQuery_(result, columns[i])
+      }
+      rows += 1
+    }
+
+    return if (rows == 0) {
+      null
+    } else {
+      val cols = arrayOfNulls<VListColumn>(query_cnt)
+      for (i in cols.indices) {
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1944]: add empty line

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt (hunk 4 above)
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1897]: No for `var orderbuf = ""`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 9]
```diff
@@ -49,7 +49,9 @@ import org.jetbrains.exposed.sql.Column
 import org.jetbrains.exposed.sql.Op
 import org.jetbrains.exposed.sql.SortOrder
 import org.jetbrains.exposed.sql.Table
+import org.jetbrains.exposed.sql.Transaction
 import org.jetbrains.exposed.sql.select
+import org.jetbrains.exposed.sql.transactions.transaction
 import org.kopi.galite.list.VListColumn
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:55]: move this import under org.kopi list

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 10]
```diff
@@ -1835,9 +1839,11 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
   }
 
   /**
-   * Warning, you should use this method under a protected statement
+   * Builds the query dialog that shows the list of data rows from database.
+   *
+   * Warning, you should use this method inside a transaction
    */
-  fun buildQueryDialog(): VListDialog? {
+  fun Transaction.buildQueryDialog(): VListDialog? {
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1846]: remove "Transaction."

**Review by h-haddad — APPROVED** [review]

**Pull request merged**